### PR TITLE
Create Tags

### PR DESCRIPTION
**Pull request by AhmedKamal20**
#6

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -2,9 +2,11 @@ class Post < ApplicationRecord
...
   validates_uniqueness_of :permalink_url, :fid
 
-  belongs_to :user, counter_cache: true
-  has_many :comments
-  has_many :reactions
+  belongs_to :user,    counter_cache: true
+  has_many :comments,  dependent: :destroy
+  has_many :reactions, dependent: :destroy
+  has_many :post_tags, dependent: :destroy
+  has_many :tags,      through: :post_tags
```
Comment (emad-elsaid) [inline app/models/post.rb]: humm, can we give up the vertical alignment? it's unnecessary

File: app/models/tag.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Tag < ApplicationRecord
+  validates_presence_of :title
+  validates_uniqueness_of :title, case_sensitive: false
```
Comment (emad-elsaid) [inline app/models/tag.rb]: I didn't know about this `case_sensitive` option, thanks :heart:

File: app/services/import_tag_service.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class ImportTagService < ApplicationService
+  def initialize(post)
+   @post = post
+  end
+
+  def call
+    tags = post.message.scan(/(?:\s|^)(?:#(?!\d+(?:\s|$)))(\w+)(?=\s|$)/i).flatten.map(&:downcase)
```
Comment (emad-elsaid) [inline app/services/import_tag_service.rb]: the `\w` matcher will match only the english characters correct? should we also match arabic characters or should we move it to future changes after we're done with this?

**Comment by emad-elsaid** [comment]
alright, ping me when you feels we can merge that, we can even have that in, then add the changes to controllers and views

File: app/models/post.rb (hunk 1 above)
Comment (AhmedKamal20) [reply app/models/post.rb]: yeah, of course we can

File: app/services/import_tag_service.rb (hunk 3 above)
Comment (AhmedKamal20) [reply app/services/import_tag_service.rb]: yeah , this regex is a black box for me right now :blush:  , i just used it from Stackoverflow , so i think we can check it again later

**Comment by AhmedKamal20** [comment]
I'm Done , tell me what you think, or merge it if it's ready

File: app/models/application_record.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 class ApplicationRecord < ActiveRecord::Base
   self.abstract_class = true
 
-  default_scope { order(created_at: :desc) }
+  default_scope { order(created_at: :desc) } # TODO : Remove this.
```
Comment (emad-elsaid) [inline app/models/application_record.rb]: I removed it already and replaced it with a service to order the model, can you rebase the branch to master? as also there is a conflict.

File: app/controllers/tags_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class TagsController < ApplicationController
+  include PageableConcern
+  before_action :set_tag, only: :show
+
+  # GET /tags
+  def index
+    @tags = paginate Tag.order_by_most_posts, 10000
```
Comment (emad-elsaid) [inline app/controllers/tags_controller.rb]: I also removed the "order_by_most" method, for ordering use the order service, I added it yesterday.

File: app/models/tag.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+  has_many :post_tags, dependent: :destroy
+  has_many :posts, through: :post_tags
+
+  scope :order_by_most_posts, -> { unscoped.order(posts_count: :desc) }
```
Comment (emad-elsaid) [inline app/models/tag.rb]: I removed a similar scope from the user model, I separated this behaviour to a service to make it work to multiple models, that makes it cleaner from my point of view

File: app/views/layouts/_navbar.html.haml [hunk 7]
```diff
@@ -1,10 +1,10 @@
 %nav.navbar.navbar-dark.bg-primary.navbar-expand-lg
   .container
-    %a.navbar-brand{:href => "#"} Egyptian Geeks
+    %a.navbar-brand{:href => "#{root_path}"} Egyptian Geeks
```
Comment (emad-elsaid) [inline app/views/layouts/_navbar.html.haml]: this can be written as
```ruby
= link_to "Egyptian Geeks", root_path, class: 'navbar-brand'
```

**Comment by AhmedKamal20** [comment]
Done

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -17,7 +17,8 @@ def index
 
   # GET /users/1
   def show
-    @posts = paginate @user.posts
+    @posts = OrderService.call(@user.posts)
```
Comment (emad-elsaid) [inline app/controllers/users_controller.rb]: yeaaaah I forgot that 😅

**Pull request merged**